src/core/ext/transport/chttp2/transport/chttp2_transport.cc

```diff
@@ -1198,10 +1198,30 @@ void grpc_chttp2_add_incoming_goaway(grpc_chttp2_transport* t,
 
 static void maybe_start_some_streams(grpc_chttp2_transport* t) {
   grpc_chttp2_stream* s;
+  /* cancel out streams that will never be started */
+  if (t->goaway_error != GRPC_ERROR_NONE) {
+    while (grpc_chttp2_list_pop_waiting_for_concurrency(t, &s)) {
+      grpc_chttp2_cancel_stream(
+          t, s,
+          grpc_error_set_int(
+              GRPC_ERROR_CREATE_FROM_STATIC_STRING("GOAWAY received"),
+              GRPC_ERROR_INT_GRPC_STATUS, GRPC_STATUS_UNAVAILABLE));
+    }
+    return;
+  }
+  if (t->next_stream_id >= MAX_CLIENT_STREAM_ID) {
```

Comment:
Is the logic correct here?
`t->next_stream_id` may be incremented in the next loop and you are not cancelling the streams any more.

Reply:
well, we won't cancel those streams in the same function anymore, but we call maybe_start_some_streams again when streams get cancelled

Reply:
hmm, wait it checks whether s->id is zero, and only does it when it is not zero.

Reply:
good catch. i'll fix it

Reply:
Since we were able to allocate a stream id, there would be a stream s->id that is not zero that would be removed in the future, so there is no correctness issue. But, it is better to cancel out streams immediately, I guess.